src/foam/lib/parse/Grammar.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+ foam.CLASS({
+  package: 'foam.lib.parse',
+  name: 'Grammar',
+
+  javaImports: [
+    'foam.lib.parse.Parser',
+    'foam.lib.parse.StringPStream',
+    'java.util.Map',
+    'foam.lib.parse.SymbolParser'
+  ],
+
+  requires: [
+    'foam.parse.ParserWithAction'
+  ],
+
+  properties: [
+    {
+      class: 'Map',
+      name: 'symbols',
+      javaFactory: 'return new java.util.HashMap<String, foam.lib.parse.Parser>();'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'addSymbol',
+      args: [
+        { name: 'symName', type: 'String' },
+        { name: 'symParser', type: 'Parser' }
+      ],
+      javaCode: `
+      getSymbols().put(symName, symParser);
+      `
+    },
+    {
+      name: 'parseString',
+      type: 'foam.lib.parse.PStream',
+      args: [
+        { name: 'str', type: 'String' },
+        { name: 'optName', type: 'String' },
+        { name: 'x',  javaType: 'foam.lib.parse.ParserContext' }
+      ],
+      javaCode: `
+if ( optName.equals("") ) optName = "START";
+ps.setString(str);
+PStream temp = ps.apply((Parser)getSymbols().get(optName), x);
+return temp;
+      `
+    },
+    {
+      name: 'sym',
+      args: [ { name: 'name', type: 'String' } ],
+      type: 'foam.lib.parse.Parser',
+      javaCode: `
+      SymbolParser symParser = new SymbolParser();
+      symParser.setSymbolName(name);
+      symParser.setSymbols(getSymbols());
+      return symParser;
+      `
+    },
+    {
+      name:'addAction',
+      args: [
+        { name: 'name', type: 'String'},
+        { name: 'action', type: 'foam.lib.parse.GrammarAction' }
+      ],
+      javaCode: `
+Map symbols = getSymbols();
+GrammarParser grParser = new GrammarParser();
+grParser.setAction(action);
+grParser.setParser((Parser) symbols.get(name));
+symbols.put(name, grParser);
+      `
+    }
+  ],
+
+  axioms: [
+    {
+      name: 'javaExtras',
+      buildJavaClass: function(cls) {
+        cls.extras.push(`
+        public StringPStream ps = new StringPStream();
```

Comment:
Java code needs to be re-entrant, unlike JS because Java is multi-threaded. Better to either make this thread-local of just create a new one in the parseString method.